crates/server/src/store/sqlite/mod.rs

```diff
@@ -881,6 +874,93 @@ impl SqliteAuthority {
 
         Ok(updated)
     }
+
+    #[cfg(feature = "__dnssec")]
+    async fn authorized_sig0(
+        &self,
+        update_message: &MessageRequest,
+        sig0: &Record,
+    ) -> UpdateResult<()> {
+        debug!("authorizing with: {sig0:?}");
+
+        let Some(sig0) = sig0.data().as_dnssec().and_then(DNSSECRData::as_sig) else {
+            warn!(
+                "no sig0 matched registered records: id {}",
+                update_message.id()
+            );
+            return Err(ResponseCode::Refused);
+        };
+
+        let name = LowerName::from(sig0.signer_name());
+        let Continue(Ok(keys)) = self
+            .lookup(&name, RecordType::KEY, LookupOptions::default())
+            .await
+        else {
+            warn!("no sig0 key name matched: id {}", update_message.id());
+            return Err(ResponseCode::Refused);
+        };
+
+        debug!("found keys {keys:?}");
+        let verified = keys
+            .iter()
+            .filter_map(|rr_set| rr_set.data().as_dnssec().and_then(DNSSECRData::as_key))
+            .any(
+                |key| match key.verify_message(update_message, sig0.sig(), sig0) {
+                    Ok(_) => {
+                        info!("verified sig: {sig0:?} with key: {key:?}");
+                        true
+                    }
+                    Err(_) => {
+                        debug!("did not verify sig: {sig0:?} with key: {key:?}");
+                        false
+                    }
+                },
+            );
+        match verified {
+            true => Ok(()),
+            false => {
+                warn!("invalid sig0 signature: id {}", update_message.id());
+                Err(ResponseCode::Refused)
+            }
+        }
+    }
+
+    #[cfg(feature = "__dnssec")]
+    async fn authorized_tsig(
+        &self,
+        update_message: &MessageRequest,
+        tsig: &Record,
+    ) -> UpdateResult<()> {
+        debug!("authorizing with: {tsig:?}");
+        let Some(signer) = self
+            .tsig_signers
+            .iter()
+            .find(|signer| signer.signer_name() == tsig.name())
+        else {
+            warn!("no TSIG key name matched: id {}", update_message.id());
+            return Err(ResponseCode::Refused);
+        };
+
+        let Ok((_, _, range)) = signer.verify_message_byte(
+            update_message.to_bytes().unwrap_or_default().as_ref(),
```

Comment:
Ah, this is where the compression issue is coming from. The `to_bytes()` call is re-encoding the message, but it looks like MAC verification needs to start with the original message off the wire. The original datagram is only present much earlier, in `handle_request()`. We'll need to use something like `DnsResponse`, which holds both the parsed message and the original encoding.

Reply:
Thanks for digging in. That matches what I assumed was happening and lines up with what @djc suggested to me in DMs, that we need a way to hold onto the original encoding :+1: 

I'm tied up for the next day or two but would be happy to try and start looking at what needs to shift to solve that issue if you aren't keen on doing it yourself before then. LMK

Reply:
I think this also applies to the pre-existing SIG0 code (?). I believe it feeds the already parsed `MessageRequest` through a `BinEncoder` to get [its TBS bytes](https://github.com/hickory-dns/hickory-dns/blob/375792ebd6d34de12cacd0c21b9fe8b7d0724365/crates/proto/src/dnssec/tbs.rs#L237).